data_diff/database.py

```diff
@@ -110,6 +112,38 @@ def md5_to_int(self, s: str) -> str:
         "Provide SQL for computing md5 and returning an int"
         ...
 
+    @abstractmethod
+    def _query(self, sql_code: str) -> list:
+        "Send query to database and return result"
+        ...
+
+
+class ThreadedDatabase(Database):
+    """Access the database through singleton threads.
+
+    Used for database connectors that do not support sharing their connection between different threads.
+    """
+
+    def __init__(self, thread_count=1):
+        self._queue = ThreadPoolExecutor(thread_count, initializer=self.set_conn)
+        self.thread_local = threading.local()
+
+    def set_conn(self):
+        assert not hasattr(self.thread_local, "conn")
+        self.thread_local.conn = self.create_connection()
+
+    def _query(self, sql_code: str):
+        r = self._queue.submit(self._query_in_worker, sql_code)
+        return r.result()
+
+    def _query_in_worker(self, sql_code: str):
+        "This method runs in a worker thread"
+        return _query_conn(self.thread_local.conn, sql_code)

```

Comment:
I'm not too familiar with error handling in threads in Python–if an exception is raised in the thread pool, is it propagated to kill the main thread? 👂🏻 I think it should. 

I ask because e.g. in Ruby/Go, that's not the case

Reply:
That's how it's supposed to work. I will verify.